### PR TITLE
[FIX] onboarding: unlink progress when company_id is unlinked

### PR DESCRIPTION
**Pull request by stpa-odoo**
### Steps to reproduce issue:

1. Download Accounting (or Sales or just Invoices)
2. Open Accounting
3. Create Company 2 and switch to it
4. Open Accounting again
5. Remove Company 2
6. Open Accounting
7. You receive an error:

>     ValueError: Expected singleton: onboarding.progress(1, 2)

### Explanation:

In `_compute_current_progress`, the filter to get `current_progress_id` accepts `onboarding.progress` with both the company in which the user is at the moment or no company at all. https://github.com/odoo/odoo/blob/77f9ff50db3cdb88397d0b1cc7042c772d0d417b/addons/onboarding/models/onboarding_onboarding.py#L55-L69 This is due to the fact that some onboardings are not related to a company while others are. After that, getting the `onboarding_state` will trigger an `ensure_one` check.

In the current case, an `onboarding.progress` with a `company_id` is not deleted once the company is deleted. Therefore the value becomes `False`. The `ensure_one` that comes after will throw an error because of it.

### Suggested fix:

`ondelete` decorator would require the module to be upgraded. For this issue, it is preferable to have a fix that is automatically applied. The added method simulates the `cascade` effect of `ondelete` and does not require the module upgrade.

opw-3762382

### CODE REVIEW

**Review by nle-odoo — APPROVED** [review]
looks good to me

> `ondelete` decorator is not accepted in stable versions. The added method simulates its `cascade` behaviour.

I thought  it was acceptable. Usually change to database are not wanted (because they can create issue for migration / custom code / custom modules / people that don't update / ...), but in this particular case (adding `ondelete="cascade"`) I think it didn't cause any of these issue  not cause an issue to people as far as I can tell (just that people that have installed/updated `onboarding` module after the `ondelete` was changed will have it up to date, and other not).

But maybe it's like this now or I remember wrongly, in any case both are ok to me depending on what is acceptable (and it would be better to have the `ondelete="cascade"` in master in any case :+1: )

side note:

I had only read the explanation on the commit and not the updated on on the PR message, if it's important that people get the fix without updating the module, I guess it could make sense.

File: addons/onboarding/models/__init__.py [hunk 1]
```diff
@@ -5,3 +5,4 @@
 from . import onboarding_onboarding_step
 from . import onboarding_progress
 from . import onboarding_progress_step
+from . import company
```
Comment (nle-odoo) [inline addons/onboarding/models/__init__.py:8]: small detail: since it's currently alphabetical we could keep the order (and while we are at it, it's better to call  it res_company.py, there is 6 cases where the file is called company.py and 168 where it's called res_company.py)

**Review by flch-odoo — APPROVED** [review]
Thank you for the fix @stpa-odoo and @nle-odoo ,

It was indeed an overlook. Thankfully, `ondelete='cascade'` is present in the progress steps so we don't have to handle these.

I thought the `onboarding_progress_onboarding_company_uniq` index would prevent nullifying the `company_id` value in the table though... Maybe the client don't have it somehow?

**Comment by stpa-odoo** [comment]
> I thought the onboarding_progress_onboarding_company_uniq index would prevent nullifying the company_id value in the table though... Maybe some clients don't have it somehow?

I don't think the index is working. The issue is reproducible on runbot, demo, trials and locally. My guess is that the index coalesces values with a 0, but `company_id` in progresses with no companies is NULL, not 0.

Also, `ondelete` will be added in master since the need to upgrade is only a problem for released versions.

**Comment by nle-odoo** [comment]
robodoo delegate+

> I thought the onboarding_progress_onboarding_company_uniq index would prevent nullifying the company_id value in the table though... Maybe some clients don't have it somehow?

With the constraint there should be max one NULL company per onboarding.onboarding record (since the constraint is on `(onboarding_id, COALESCE(company_id, 0))`). So I guess there would be a unicity error only if we delete 2 companies.

**Comment by flch-odoo** [comment]
You are both correct about this index, thank you for the replies. We're still good with the fix (and switch to `ondelete` in master) :)

**Comment by stpa-odoo** [comment]
robodoo r+

**Comment by tde-banana-odoo** [comment]
Wait, what ? Why aren't you adding ondelete='cascade' ?

**Comment by tde-banana-odoo** [comment]
@robodoo r- (waiting for some discussion)

**Comment by stpa-odoo** [comment]
@tde-banana-odoo The idea is to fix the issue without the need for a manual operation on an existing database.
This issue can block the access to apps as a whole. In later versions (namely 17.1), it causes an internal server error. Because of this, I would prefer having the fix applied automatically.

**Comment by flch-odoo** [comment]
I've checked and found:
1. Only using "ondelete='cascade'" does not lead to an update of the FK, (nor does it trigger a cascade from the ORM when we unlink the company),
2. Even "upgrading" the `onboarding` module from `Apps` after this change does not help, so we'll need an upgrade script in master,
3. The current fix will prevent the situation from arising but not fix impacted DBs, so we'll need something else from 16.4 too... Perhaps within `_search_or_create_progress`, called when we try to load a banner, we could check for this unfortunate situation and clean (remove progress with company=NULL if there are others with company set)?

**Comment by flch-odoo** [comment]
Likely last update for today:
Let's please
1. Forget the clean from the code
2. Add the `ondelete` from `16.4` (together with the current solution), as it does not appear to break anything.

The question of having the foreign key updated remains, I'll talk with upgrade tomorrow.

**Comment by stpa-odoo** [comment]
1. What does "FK" mean?
2. After testing it on my side, upgrading `onboarding` from the apps does work, the progress is succesfully deleted in the database.
3. `onboarding.onbarding` has a boolean `is_per_company` that could be useful for this.

**Comment by nle-odoo** [comment]
Hi,

> The current fix will prevent the situation from arising but not fix impacted DBs, so we'll need something else from 16.4 too... Perhaps within _search_or_create_progress, called when we try to load a banner, we could check for this unfortunate situation and clean (remove progress with company=NULL if there are others with company set)?

as a bandage for stable in addition to `ondelete="cascade"`, couldn't we fallback to first record until this is fixed? eg.

```diff
--- a/addons/onboarding/models/onboarding_onboarding.py
+++ b/addons/onboarding/models/onboarding_onboarding.py
@@ -57,7 +57,8 @@ class Onboarding(models.Model):
     def _compute_current_progress(self):
         for onboarding in self:
             current_progress_id = onboarding.progress_ids.filtered(
-                lambda progress: progress.company_id.id in {False, self.env.company.id})
+                lambda progress: progress.company_id.id in {False, self.env.company.id}
+            ).sorted('company_id', reverse=True)[:1]
             if current_progress_id:
```
it seems to allow to open the onboarding at least, and new instance / updated instances will not have this issue.

> What does "FK" mean?

foreign key, the company_id column in onboarding_progress is a foreign key to the id column of table res_company

> After testing it on my side, upgrading onboarding from the apps does work, the progress is succesfully deleted in the database.

**Comment by flch-odoo** [comment]
Today I learned that my IDE's DB explorer keeps a cache of table definitions and this is why I didn't see it, now I know better.
(Also, the existing progress record with NULL company was still there (to not help me^^) but this is not what we should solve for, so never mind.)

@nle-odoo Your suggestion would prevent/delay impacted users from contacting support to get their records fixed, and we'd prefer that they did over making an migration script.

So, to summarize, we can add the `ondelete="cascade"` in saas-16.4, remove the `@api.ondelete` in master's forward port, and rely on support to fix remaining DBs (as they're doing now). Am I missing anything?

Thank you both for the top-notch follow-up!

**Review by flch-odoo — APPROVED** [review]
@tde-banana-odoo , is this OK with you too?

File: addons/onboarding/models/res_company.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+from odoo import api, models
+
+
+class Company(models.Model):
+    _inherit = 'res.company'
+
+    @api.ondelete(at_uninstall=False)
+    def _unlink_onboarding_progress(self):
+        progress = self.env['onboarding.progress'].search([('company_id', 'in', self.ids)])
```
[the remark this answers is not in the document]
Comment (flch-odoo) [reply addons/onboarding/models/res_company.py:12]: In standard, only `group_erp_manager` can delete companies; 
`group_system` implies it, and this `group_system` is the only group allowed any CRUD on onboarding (including progresses).

[the remark this answers is not in the document]
Comment (stpa-odoo) [reply addons/onboarding/models/res_company.py:12]: Is a sudo needed or is it impossible to give `group_erp_manager` withtout `group_system`?

[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/onboarding/models/res_company.py:12]: > In standard, only `group_erp_manager` can delete companies; `group_system` implies it, and this `group_system` is the only group allowed any CRUD on `onboarding` (including progresses).

so there is an issue with erp_manager unlinks a company (system has more rights than erp_manager)

[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/onboarding/models/res_company.py:12]: Maybe worth using '@users' decorator in the test :)

[the remark this answers is not in the document]
Comment (stpa-odoo) [reply addons/onboarding/models/res_company.py:12]: Confirmed user gets an error.

[the remark this answers is not in the document]
Comment (flch-odoo) [reply addons/onboarding/models/res_company.py:12]: Can user delete a company ?

[the remark this answers is not in the document]
Comment (stpa-odoo) [reply addons/onboarding/models/res_company.py:12]: >     You are not allowed to access 'Onboarding Progress Tracker' (onboarding.progress) records.
>
>     This operation is allowed for the following groups:
>             - Administration/Settings
>
>     Contact your administrator to request access if necessary.

[the remark this answers is not in the document]
Comment (flch-odoo) [reply addons/onboarding/models/res_company.py:12]: Of course, but users can't delete a company either. 
So we could use sudo for this as the access right to delete a company will be checked at some point, right?
Unless user is an ERP manager (could be), the test should also raise an error when it gets to deleting the company, shouldn't it?

[the remark this answers is not in the document]
Comment (stpa-odoo) [reply addons/onboarding/models/res_company.py:12]: Oh yes, the user I am refering to has `group_erp_manager`, sorry.

[the remark this answers is not in the document]
Comment (stpa-odoo) [reply addons/onboarding/models/res_company.py]: Added the sudo, everything seems to be working fine.

File: addons/onboarding/tests/test_onboarding.py [hunk 3]
```diff
@@ -215,3 +215,20 @@ def test_onboarding_step_without_onboarding(self):
         # But it can
         self.step_initially_w_o_onboarding.with_company(self.company_2).action_set_just_done()
         self.assert_onboarding_is_done(self.onboarding_3.with_company(self.company_2))
+
+    def test_remove_company_with_progress(self):
+        user = self.env['res.users'].create({
```
Comment (tde-banana-odoo) [inline addons/onboarding/tests/test_onboarding.py:220]: You may import and use ``_mail_new_test_user`` tool to create users (include all necessary keys to avoid computation and speedup user creation, we don't need confirmation emails or stuff like that in tests :) )

**Review by tde-banana-odoo — COMMENTED** [review]
Small proposal

**Comment by stpa-odoo** [comment]
So, I added the sudo and switched to `mail_new_test_user`. Do you see anything else that I could improve?

**Review by tde-banana-odoo — APPROVED** [review]
@robodoo r+
